proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,471 @@
+
+The proposal is to provide a set of CompilerServices apis implemented as intrinsics (in most cases) which will allow the use of more complex constants. The intention is to permit more complex constant forms for ReadOnlySpan<T> access, constant access, and through IL compiler magic, add new forms of array initialization that are not currently supported through the InitializeArray pattern.
+
+### Definition of types which can be represented as constants
+- Must be a struct
+- Generic structures are permitted, and only those generic type parameters which contribute to field layout shall have impact on whether the type can be represented as a constant.
+- The struct may have any form of layout: sequential layout, auto layout, or explicit layout.
+- Without pointers of any form (No object reference, IntPtr, UIntPtr, pointer, function pointer, or TypedReference)
+- The type must be attributed with either WellKnownLayoutTypeAttribute or WellKnownLayoutVectorTypeAttribute
+- All fields must also be of a type which can be represented as a constant
+
+### Well known type layout
+Given that IL binaries may be loaded on many different platforms with different type layout rules, constant data must be represented in a consistent manner across all of them. For consistent type layout, types shall be laid out in precisely a sequential manner as described in metadata, with *no* packing between any fields. For instance, this struct will utilize 9 bytes when stored. In addition, if the type uses explicit layout, even if there are overlapping fields, those fields are not to be overlapped in the WellKnownLayout, the layout of the final form will be based on copying out of the data stored in the PE file in metadata order into the runtime layout. An empty structure will consume 0 bytes of storage, and thus while it may be used as a field within a constant, is not itself permitted to be used to instantiate CreateSpan<T> or LoadConstant<T>
+
+```csharp
+struct NonAlignedStruct
+{
+    byte b;
+    double d;
+}
+```
+
+### As part of this proposal attributes will be placed on various types
+The WellKnownLayoutTypeAttribute will be placed onto
+
+System.Byte
+System.SByte
+System.Boolean
+System.Int16
+System.UInt16
+System.Char
+System.Int32
+System.UInt32
+System.Float
+System.Int64
+System.UInt64
+System.Double
+
+System.Numerics.Matrix3x2
+System.Numerics.Matrix4x4
+System.Numerics.Plane
+System.Numerics.Quaternion
+System.Numerics.Vector2
+System.Numerics.Vector3
+System.Numerics.Vector4
+
+The WellKnownLayoutVectorTypeAttribute will be placed onto
+System.Runtime.Intrinsics.Vector64<T>
+System.Runtime.Intrinsics.Vector128<T>
+System.Runtime.Intrinsics.Vector256<T>
+
+### RuntimeHelpers apis
+
+```csharp
+class RuntimeHelpers
+{
+    // Placing this attribute on a type indicates that it can be stored and loaded using the CreateSpan<T>, LoadConstant<T> and InitializeArray helper functions
+    public WellKnownLayoutTypeAttribute : Attribute
```

Comment:
We never have attributes as nested types.

Reply:
Shouldn't this attribute be synthetic and encoded as 0x18 in the type flags?